api/valor_api/backend/query/filtering.py

```diff
@@ -1,430 +1,379 @@
 import operator
-from typing import Callable
 
-from sqlalchemy import TIMESTAMP, Boolean, Float, and_, cast, func, not_, or_
+from geoalchemy2.functions import ST_Area, ST_Count, ST_GeomFromGeoJSON
+from sqlalchemy import (
+    CTE,
+    TIMESTAMP,
+    BinaryExpression,
+    Boolean,
+    Float,
+    Integer,
+    and_,
+    cast,
+    func,
+    not_,
+    or_,
+    select,
+)
 from sqlalchemy.dialects.postgresql import INTERVAL, TEXT
-from sqlalchemy.sql.elements import BinaryExpression, ColumnElement
 
-from valor_api import enums
 from valor_api.backend.models import (
     Annotation,
     Dataset,
     Datum,
+    Embedding,
+    GroundTruth,
     Label,
     Model,
     Prediction,
 )
-from valor_api.backend.query.types import TableTypeAlias
-from valor_api.schemas import (
-    BooleanFilter,
-    DateTimeFilter,
-    Duration,
-    Filter,
-    GeospatialFilter,
-    NumericFilter,
-    StringFilter,
-    Time,
+from valor_api.backend.query.types import LabelSourceAlias, TableTypeAlias
+from valor_api.schemas.filters import AdvancedFilter as Filter
+from valor_api.schemas.filters import (
+    NArgFunction,
+    OneArgFunction,
+    Symbol,
+    TwoArgFunction,
+    Value,
+)
+from valor_api.schemas.geometry import (
+    Box,
+    LineString,
+    MultiLineString,
+    MultiPoint,
+    MultiPolygon,
+    Point,
+    Polygon,
 )
 
-
-def _get_boolean_op(opstr) -> Callable:
-    """Returns function if operator is valid for boolean comparison."""
-    ops = {
-        "==": operator.eq,
-        "!=": operator.ne,
-    }
-    if opstr not in ops:
-        raise ValueError(f"invalid boolean comparison operator `{opstr}`")
-    return ops[opstr]
-
-
-def _get_string_op(opstr) -> Callable:
-    """Returns function if operator is valid for string comparison."""
-    ops = {
-        "==": operator.eq,
-        "!=": operator.ne,
-    }
-    if opstr not in ops:
-        raise ValueError(f"invalid string comparison operator `{opstr}`")
-    return ops[opstr]
-
-
-def _get_numeric_op(opstr) -> Callable:
-    """Returns function if operator is valid for numeric comparison."""
-    ops = {
-        ">": operator.gt,
-        "<": operator.lt,
-        ">=": operator.ge,
-        "<=": operator.le,
-        "==": operator.eq,
-        "!=": operator.ne,
-    }
-    if opstr not in ops:
-        raise ValueError(f"invalid numeric comparison operator `{opstr}`")
-    return ops[opstr]
-
-
-def _get_spatial_op(opstr) -> Callable:
-    """Returns function if operator is valid for spatial comparison."""
-    ops = {
-        "intersect": lambda lhs, rhs: func.ST_Intersects(lhs, rhs),
-        "inside": lambda lhs, rhs: func.ST_Covers(rhs, lhs),
-        "outside": lambda lhs, rhs: not_(func.ST_Covers(rhs, lhs)),
+opstr_to_operator = {
```

Comment:
nit: add short docstrings above these dictionaries describing what they're used for?